tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -8340,6 +8320,108 @@ Sema::TemplateDeductionResult HLSLExternalSource::DeduceTemplateArgumentsForHLSL
       continue;
     }
 
+    // Currently only intrinsic we allow for explicit template arguments are
+    // for Load return types for ByteAddressBuffer/RWByteAddressBuffer
+    // TODO: handle template arguments for future intrinsics in a more natural way
+
+    // Check Explicit template arguments
+    LPCSTR intrinsicName = (*cursor)->pArgs[0].pName;
+    if (ExplicitTemplateArgs) {
```

Comment:
So much code and so many diagnostic branches... is it possible to simplify?  Perhaps:
```
bool Is2018 = getSema()->getLangOpts().HLSLVersion >= 2018;
bool IsBAB = functionParentRecord == m_objectTypeDecls[AR_OBJECT_BYTEADDRESS_BUFFER] ||
              functionParentRecord == m_objectTypeDecls[AR_OBJECT_RWBYTEADDRESS_BUFFER];
bool IsBABLoad = bIsBAB && intrinsicOp == (UINT)IntrinsicOp::MOP_Load;
//bool IsBABStore = bIsBAB && intrinsicOp == (UINT)IntrinsicOp::MOP_Store; // do we need this?
if (ExplicitTemplateArgs && ExplicitTemplateArgs->size() > 0) {
  bool IsLegal = false;
  SourceLocation Loc = ExplicitTemplateArgs->getLAngleLoc();
  const char *TemplateSupportText = (Is2018 && IsBABLoad) ? // or IsBABStore?
          "One scalar or vector type argument up to 16 bytes in size" : 
          "No template arguments";
  if (IsBABLoad && Is2018 && ExplicitTemplateArgs->size() == 1) { // or IsBABStore?
    Loc = (*ExplicitTemplateArgs)[0].getLocation();
    QualType explicitType = (*ExplicitTemplateArgs)[0].getArgument().getAsType();
    ArTypeObjectKind explicitKind = GetTypeObjectKind(explicitType);
    if (explicitKind == AR_TOBJ_BASIC || explicitKind == AR_TOBJ_VECTOR) {
      bool IsLegal = GET_BASIC_BITS(GetTypeElementKind(explicitType)) != BPROP_BITS64) ||
                     GetNumElements(explicitType) <= 2;
    }
    if (IsLegal) {
      if (IsBABLoad)
        argTypes[0] = explicitType;
      //else if (IsBABStore)
      //  argTypes[2] = explicitType;
    }
  }
  if (!IsLegal) {
    getSema()->Diag(Loc, diag::err_hlsl_unsupported_template_for_intrinsic)
                    << intrinsicName << TemplateSupportText;
    return Sema::TemplateDeductionResult::TDK_Invalid;
  }
} else if (IsBABStore && !Is2018) {
  // Prior to HLSL 2018, Store operation for ByteAddressBuffer only stored uint.
  if (GetNumElements(argTypes[2]) != 1) {
    getSema()->Diag(Args[1]->getLocStart(), diag::err_ovl_no_viable_member_function_in_call)
      << intrinsicName;
    return Sema::TemplateDeductionResult::TDK_Invalid;
  }
  argTypes[2] = getSema()->getASTContext().getIntTypeForBitwidth(32, /*signed*/ false);
}
```